etag/middleware.go

```diff
@@ -14,31 +15,53 @@ const (
 	IfNoneMatchHeaderKey = "If-None-Match"
 )
 
+var (
+	errDoubleWrite = errors.New("Write called twice with ETag Middleware")
+)
+
 type etagResponseWriter struct {
 	http.ResponseWriter
-	req  *http.Request
-	code int
+	req         *http.Request
+	code        int
+	wroteHeader bool
+	wroteBody   bool
 }
 
 func (erw *etagResponseWriter) Write(b []byte) (int, error) {
-	etag := etag(b)
+	etag := calculateEtag(b)
 	erw.Header().Set(ETagHeaderKey, etag)
```

Comment:
`Etag` header only makes sense when returning a success status code (< `300`). I've seen some bugs become really nasty because of this :(